fixed-point/Converter/Bonsai.py

```diff
@@ -0,0 +1,321 @@
+        self.T = readFileAsMat(os.path.join(
+            getModelDir(), "Theta"), "\t", float)
+        self.Sigma = readFileAsMat(os.path.join(
+            getModelDir(), "Sigma"), "\t", float)
+        self.Mean = readFileAsMat(os.path.join(
+            getModelDir(), "Mean"), "\t", float)
+        self.Variance = readFileAsMat(os.path.join(
+            getModelDir(), "Variance"), "\t", float)
+
+    def validateModel(self):
+        Z_m, Z_n = matShape(self.Z)
+        W_m, W_n = matShape(self.W)
+        V_m, V_n = matShape(self.V)
+        # If T is empty
+        if len(self.T) == 0:
+            T_m, T_n = 0, Z_m
+        else:
+            T_m, T_n = matShape(self.T)
+        Sigma_m, Sigma_n = matShape(self.Sigma)
+        Mean_m, Mean_n = matShape(self.Mean)
+        Variance_m, Variance_n = matShape(self.Variance)
+
+        assert Z_n == Mean_m == Variance_m
+        assert Z_m == W_n == V_n == T_n
+        assert W_m == V_m
+        assert Sigma_m == Sigma_n == 1
+        assert Mean_n == Variance_n == 1
+
+    # Restructure the W and V matrix to be node ID major instead of class ID major
+    def rearrangeWandV(self, mat):
+        matNew = [[] for _ in range(len(mat))]
+        for i in range(self.numClasses):
+            for j in range(self.totalNodes):
+                matNew[i + j * self.numClasses] = mat[i * self.totalNodes + j]
+        return matNew
+
+    # Note: Following computations assume that the tree is always complete (which is a property of Bonsai algorithm)
+    def computeVars(self):
+        self.D = len(self.Z[0])
+        self.d = len(self.Z)
+        self.internalNodes = len(self.T)
+        self.depth = int(math.log2(self.internalNodes + 1))
+        self.totalNodes = 2 ** (self.depth + 1) - 1
+        self.numClasses = len(self.W) // self.totalNodes
+
+    def formatModel(self):
+        # Manually set mean and variance of bias to 0 and 1 respectively
+        self.Mean[len(self.Mean) - 1] = [0]
+        self.Variance[len(self.Variance) - 1] = [1]
+
+        sigma = self.Sigma[0][0]
+
+        self.computeVars()
+
+        # Precompute some values to speedup prediction
+        # Precompute V*sigma and Z/d
+        self.V = [[x * sigma for x in y] for y in self.V]
+        self.Z = [[x / self.d for x in y] for y in self.Z]
+
+        # Precompute Z*(X-m)/v by absorbing m, v into Z
+        self.Z = [[x / v[0] for x, v in zip(y, self.Variance)] for y in self.Z]
+
+        self.mean = matMul(self.Z, self.Mean)
+        assert len(self.mean[0]) == 1
+        self.mean = [x[0] for x in self.mean]
+
+        # Restructure W and V
+        self.W = self.rearrangeWandV(self.W)
+        self.V = self.rearrangeWandV(self.V)
+
+    def verifyModel(self):
+        # If T is empty, it is replaced with a matrix containing 2 elements (which is never used during prediction)
+        if len(self.T) == 0:
+            print("Warning: Empty matrix T\n")
+            self.T = [[-0.000001, 0.000001]]
+
```

Comment:
Why do you need the two elements?

Reply:
So that the input for the compiler can clearly distinguish the range of elements in T.
This will reflect in the input.txt which is generated.